### PR TITLE
BUG: increase precision

### LINKED ISSUE #414: DCM Seg: Rounding of orientation matrix when storing seg
It seems that DCMQI rounds values at the 5th decimal position, if storing DCM Segs.

We had this issues now with some data sets (see https://phabricator.mitk.org/T28127), where the rounding became obvious.
E.g.
Stored as nrrd:
`space directions: (0.68359375,0,0) (0,0.68359375,0) (0,0,1)`
Stored as DCM Seg (via DCMQI):
`space directions: (0.68359380000000003,0,0) (0,0.68359380000000003,0) (0,0,1)`

Before we just lower the precision of consistency checks on our side, we thought that it might make sense to check that DCMQI provides the highest possible fidelity when storing the data. That's why we wanted to point out that finding here and discuss it.

What could/is the reason for that rounding? Any idea?

### PR DESCRIPTION
**Pull request by fedorov**
resolves #414

### CODE REVIEW

File: libsrc/Helper.cpp [hunk 1]
```diff
@@ -103,9 +103,32 @@ namespace dcmqi {
...
+     - exponent (2 OR 3 (Win))
+
+    ostringstream sstream;
+    sstream << setprecision(8) << scientific << f;
+    cout << setprecision(8) << scientific << f << endl;
+    */
     ostringstream sstream;
-    sstream << scientific << f;
-    return sstream.str();
+    sstream << fixed << setprecision(15) << f;
```
Comment (rfloca) [inline libsrc/Helper.cpp:121]: sstream.imbue(std::locale::classic());
before using the sstream.

File: libsrc/Helper.cpp [hunk 2]
```diff
@@ -103,9 +103,32 @@ namespace dcmqi {
...
+    string formatted_f = sstream.str();
+    size_t dot_position = formatted_f.find('.');
```
Comment (rfloca) [inline libsrc/Helper.cpp:123]: You can move thin into the if statement as it is only used in that scope.

File: libsrc/Helper.cpp [hunk 3]
```diff
@@ -103,9 +103,32 @@ namespace dcmqi {
...
+    if(formatted_f.length()>16){
+      if(dot_position == string::npos || dot_position > 15){
+        cerr << "ERROR: Failed to convert " << f << " to DS VR!" << endl;
+      } else {
+        formatted_f = formatted_f.substr(0,16);
+      }
+    }
```
Comment (rfloca) [inline libsrc/Helper.cpp:130]: Should not be needed, if everything works according to standard. But it doesn't hurt, if you have no very high performance requirements.

**Review by rfloca — CHANGES_REQUESTED** [review]

**Comment by kislinsk** [comment]
Here's my shot: http://cpp.sh/72ipr

To squeeze every little possible bit of precision out of it, you could print both scientific and fixed notation to strings and compare their lengths. If the scientific notation fits, use it. Fixed notation otherwise. The numeric_limits::max_digits10 ensures that both notations represent the maximum possible precision.

**Comment by kislinsk** [comment]
Nevermind, I just noticed that DCMQI only used float so it makes things easier in the context of 16 characters. The standard library provides useful constants for serializing floating point numbers with maximum precision and for float (with all the extras) it always fits even into less than 16 characters. I will create a pull request for review.

**Comment by fedorov** [comment]
replaced by #416

**Pull request closed without merging**